File: goblin-api/src/main/java/org/goblinframework/api/rpc/ServiceRetries.java
package org.goblinframework.api.rpc;

import java.lang.annotation.*;

@Documented
@Target({ElementType.TYPE, ElementType.METHOD, ElementType.FIELD})
@Retention(RetentionPolicy.RUNTIME)
public @interface ServiceRetries {

  int retries() default 0;

  boolean enable() default true;

}

File: goblin-rpc/src/main/java/org/goblinframework/rpc/service/_ServiceRetries.kt
package org.goblinframework.rpc.service

import org.goblinframework.api.rpc.ServiceRetries
import java.lang.reflect.Method

fun calculateServiceRetries(interfaceClass: Class<*>): Int {
  check(interfaceClass.isInterface)
  val annotation = interfaceClass.getAnnotation(ServiceRetries::class.java)
  val retries = calculateServiceRetries(annotation)
  return retries ?: 0
}

fun calculateServiceRetries(method: Method): Int? {
  val serviceRetries = method.getAnnotation(ServiceRetries::class.java)
  return calculateServiceRetries(serviceRetries)
}

fun calculateServiceRetries(annotation: ServiceRetries?): Int? {
  return annotation?.run {
    if (this.enable) {
      val retries: Int = this.retries
      if (retries < 0) null else retries
    } else {
      null
    }
  }
}